Jumpscale/clients/blockchain/tfchain/TFChainCapacity.py

```diff
@@ -0,0 +1,100 @@
+import nacl
+from Jumpscale import j
+
+from . import schemas
+
+
+class TFChainCapacity():
+    """
+    TFChainReservation contains all TF grid reservation logic.
+    """
+
+    def __init__(self, wallet):
+        self._wallet = wallet
+        self._notary_client_ = None
+        self._grid_broker_pub_key_ = None
+        self._grid_broker_addr = ''  # TODO: hardcode tfchain address of the grid broker
+        # this all key management needs to be improved
+        keypair = self._wallet.key_pair_get(self._wallet.address)
+        self._signing_key = nacl.signing.SigningKey(keypair.private_key.to_seed())
+        self._private_key = self._signing_key.to_curve25519_private_key()
+
+    @property
+    def _notary_client(self):
+        if self._notary_client_ is None:
+            c = j.clients.gedis.configure('tfnotary', host='notary.grid.tf', port=6830)
+            self._notary_client_ = c.cmds.notary_actor
+        return self._notary_client_
+
+    @property
+    def _grid_broker_pub_key(self):
+        if self._grid_broker_pub_key_ is None:
+            record = self._wallet.client.threebot.record_get('tf3bot.zaibon')
+            encoded_key = record.public_key.hash
+            vk = nacl.signing.VerifyKey(
+                str(encoded_key),
+                encoder=nacl.encoding.HexEncoder)
+            self._grid_broker_pub_key_ = vk.to_curve25519_public_key()
+        return self._grid_broker_pub_key_
+
+    def reserve_s3(self, size, email, threebot_id, duration=None):
+        reservation = j.data.schema.get(url='tfchain.reservation.s3').new(data={
+            'size': size,
+            'email': email,
+            'created': j.data.time.epoch
+        })
+        return self._process_reservation(reservation, threebot_id)
+
+    def reserve_zos_vm(self, size, email, threebot_id, duration=None):
+        reservation = j.data.schema.get(url='tfchain.reservation.zos_vm').new(data={
```

Comment:
the schema automatically validates the types I suppose?

Reply:
well that's what it supposed to do...